In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os

In [2]:
# change the font size of plot
plt.rcParams.update({'font.size': 16})

In [4]:
%%time
train = pd.read_csv('data/train.csv')
# test = pd.read_csv('data/test.csv')

Wall time: 15.8 s


In [5]:
train.memory_usage(), train.memory_usage().sum()

(Index                  128
 building_id      161728800
 meter            161728800
 timestamp        161728800
 meter_reading    161728800
 dtype: int64, 646915328)

In [5]:
#del test
#del weather_test_prepared_df

In [6]:
# read data from file:
building_prepared_df = pd.read_csv('C:\\data\\Kaggle_ASHRAE_energy\\building_prepared_df.csv')
weather_train_prepared_df = pd.read_csv('C:\\data\\Kaggle_ASHRAE_energy\\weather_train_prepared_df.csv')
# weather_test_prepared_df = pd.read_csv('C:\\data\\Kaggle_ASHRAE_energy\\weather_test_prepared_df.csv')

In [7]:
building_prepared_df.shape, weather_train_prepared_df.shape

((1449, 1467), (139773, 11))

In [8]:
train.shape

(20216100, 4)

In [9]:
building_prepared_df

site_id  building_id    0    1    2    3    4    5    6    7  ...  1455  \
0           0            0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   
1           0            1  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   
2           0            2  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   
3           0            3  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...   0.0   
4           0            4  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...   0.0   
...       ...          ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   
1444       15         1444  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   
1445       15         1445  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   
1446       15         1446  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   
1447       15         1447  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   
1448       15         1448  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   

      1456  1457  1458  1459  1460  1461  1462  1463      1464  
0      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.008173  
1      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.002786  
2      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.005822  
3      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.026754  
4      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.132985  
...    ...   ...   ...   ...   ...   ...   ...   ...       ...  
1444   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.022105  
1445   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.004590  
1446   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.012555  
1447   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.033716  
1448   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.105163  

[1449 rows x 1467 columns]

In [10]:
weather_train_prepared_df.columns = ['site_id','timestamp','airtemp','sealev','dewtemp','windsp',
                                     'North','West','South','nowind','dewair']

In [11]:
weather_train_prepared_df

site_id            timestamp   airtemp    sealev   dewtemp    windsp  \
0             0  2016-01-01 00:00:00  0.698187  0.628816  0.891410  0.000000   
1             0  2016-01-01 01:00:00  0.690415  0.634921  0.909238  0.061983   
2             0  2016-01-01 02:00:00  0.669689  0.634921  0.909238  0.000000   
3             0  2016-01-01 03:00:00  0.647668  0.633700  0.901135  0.000000   
4             0  2016-01-01 04:00:00  0.633420  0.632479  0.891410  0.107438   
...         ...                  ...       ...       ...       ...       ...   
139768       15  2016-12-31 19:00:00  0.413212  0.482295  0.437601  0.235537   
139769       15  2016-12-31 20:00:00  0.410622  0.478632  0.423015  0.318182   
139770       15  2016-12-31 21:00:00  0.410622  0.479853  0.450567  0.210744   
139771       15  2016-12-31 22:00:00  0.402850  0.485958  0.458671  0.190083   
139772       15  2016-12-31 23:00:00  0.396373  0.492063  0.476499  0.363636   

        North  West  South  nowind    dewair  
0         0.0   0.0    0.0     1.0  0.960000  
1         1.0   0.0    0.0     0.0  0.973473  
2         0.0   0.0    0.0     1.0  0.986156  
3         0.0   0.0    0.0     1.0  0.995871  
4         0.0   1.0    0.0     0.0  1.000000  
...       ...   ...    ...     ...       ...  
139768    0.0   0.0    1.0     0.0  0.893204  
139769    0.0   0.0    1.0     0.0  0.886187  
139770    0.0   0.0    1.0     0.0  0.902724  
139771    0.0   0.0    1.0     0.0  0.912916  
139772    0.0   0.0    1.0     0.0  0.928220  

[139773 rows x 11 columns]

In [12]:
building_prepared_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1449 entries, 0 to 1448
Columns: 1467 entries, site_id to 1464
dtypes: float64(1465), int64(2)
memory usage: 16.2 MB


In [13]:
weather_train_prepared_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 139773 entries, 0 to 139772
Data columns (total 11 columns):
site_id      139773 non-null int64
timestamp    139773 non-null object
airtemp      139773 non-null float64
sealev       139773 non-null float64
dewtemp      139773 non-null float64
windsp       139773 non-null float64
North        139773 non-null float64
West         139773 non-null float64
South        139773 non-null float64
nowind       139773 non-null float64
dewair       139660 non-null float64
dtypes: float64(9), int64(1), object(1)
memory usage: 11.7+ MB


In [15]:
## Function to reduce the memory usage
def reduce_mem_usage(df, verbose=True):
    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
    start_mem = df.memory_usage().sum() / 1024**2    
    for col in df.columns:
        col_type = df[col].dtypes
        if col_type in numerics:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)  
            else:
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)    
    end_mem = df.memory_usage().sum() / 1024**2
    if verbose: print('Mem. usage decreased from {:5.2f} Mb to {:5.2f} Mb ({:.1f}% reduction)'.format(start_mem, end_mem, 100 * (start_mem - end_mem) / start_mem))
    return df


In [16]:
building_prepared_df = reduce_mem_usage(building_prepared_df)
weather_train_prepared_df = reduce_mem_usage(weather_train_prepared_df)
train =  reduce_mem_usage(train)

Mem. usage decreased from 16.22 Mb to  4.05 Mb (75.0% reduction)
Mem. usage decreased from 11.73 Mb to  3.60 Mb (69.3% reduction)
Mem. usage decreased from 616.95 Mb to 289.19 Mb (53.1% reduction)


In [ ]:
merge_df = pd.merge(left = building_prepared_df, right = weather_train_prepared_df,
                   left_on='site_id', right_on='site_id')
merge_df

In [327]:
merge_df = pd.merge(left = building_prepared_df[building_prepared_df.site_id == 0], 
                    right = weather_train_prepared_df[weather_train_prepared_df.site_id == 0],
                   left_on='site_id', right_on='site_id')
merge_df

MemoryError: Unable to allocate array with shape (1465, 922320) and data type float16

### Start cleaning the Y (meter reading) data.

get the site_id and the data order from building and weather

In [17]:
# merge site_id, building_id, timestamp data from building and weather :
site_building_ts = pd.merge(left=building_prepared_df[['site_id','building_id']], 
                               right = weather_train_prepared_df[['site_id','timestamp']], 
                               left_on='site_id', right_on = 'site_id')
site_building_ts

site_id  building_id            timestamp
0               0            0  2016-01-01 00:00:00
1               0            0  2016-01-01 01:00:00
2               0            0  2016-01-01 02:00:00
3               0            0  2016-01-01 03:00:00
4               0            0  2016-01-01 04:00:00
...           ...          ...                  ...
12676161       15         1448  2016-12-31 19:00:00
12676162       15         1448  2016-12-31 20:00:00
12676163       15         1448  2016-12-31 21:00:00
12676164       15         1448  2016-12-31 22:00:00
12676165       15         1448  2016-12-31 23:00:00

[12676166 rows x 3 columns]

#### merge table
merge dataset above with meter_reading data and split into 4 dataset by different meters

In [18]:
train_set0 = pd.merge(left= site_building_ts, right = train.query('meter == 0').drop('meter',axis=1), 
                      left_on=['building_id','timestamp'], right_on = ['building_id','timestamp'])
train_set1 = pd.merge(left= site_building_ts, right = train.query('meter == 1').drop('meter',axis=1), 
                      left_on=['building_id','timestamp'], right_on = ['building_id','timestamp'])
train_set2 = pd.merge(left= site_building_ts, right = train.query('meter == 2').drop('meter',axis=1), 
                      left_on=['building_id','timestamp'], right_on = ['building_id','timestamp'])
train_set3 = pd.merge(left= site_building_ts, right = train.query('meter == 3').drop('meter',axis=1), 
                      left_on=['building_id','timestamp'], right_on = ['building_id','timestamp'])

In [23]:
# I can get the same data order from below query too, but missing site_id
# train.sort_values(by = ['building_id','meter','timestamp'])

query shows lots of zeros for meter0 reading

In [19]:
train_set0.head(30)

site_id  building_id            timestamp  meter_reading
0         0            0  2016-01-01 00:00:00            0.0
1         0            0  2016-01-01 01:00:00            0.0
2         0            0  2016-01-01 02:00:00            0.0
3         0            0  2016-01-01 03:00:00            0.0
4         0            0  2016-01-01 04:00:00            0.0
5         0            0  2016-01-01 05:00:00            0.0
6         0            0  2016-01-01 06:00:00            0.0
7         0            0  2016-01-01 07:00:00            0.0
8         0            0  2016-01-01 08:00:00            0.0
9         0            0  2016-01-01 09:00:00            0.0
10        0            0  2016-01-01 10:00:00            0.0
11        0            0  2016-01-01 11:00:00            0.0
12        0            0  2016-01-01 12:00:00            0.0
13        0            0  2016-01-01 13:00:00            0.0
14        0            0  2016-01-01 14:00:00            0.0
15        0            0  2016-01-01 15:00:00            0.0
16        0            0  2016-01-01 16:00:00            0.0
17        0            0  2016-01-01 17:00:00            0.0
18        0            0  2016-01-01 18:00:00            0.0
19        0            0  2016-01-01 19:00:00            0.0
20        0            0  2016-01-01 20:00:00            0.0
21        0            0  2016-01-01 21:00:00            0.0
22        0            0  2016-01-01 22:00:00            0.0
23        0            0  2016-01-01 23:00:00            0.0
24        0            0  2016-01-02 00:00:00            0.0
25        0            0  2016-01-02 01:00:00            0.0
26        0            0  2016-01-02 02:00:00            0.0
27        0            0  2016-01-02 03:00:00            0.0
28        0            0  2016-01-02 04:00:00            0.0
29        0            0  2016-01-02 05:00:00            0.0

below query also tells me there is a scenario that some building didn't have the meter initially, but install it sometime after. I should run that 0 meter deletion on ALL buildings and all meters.not only on some

In [15]:
train_set2.query('building_id == 745')

site_id  building_id            timestamp  meter_reading
0           6          745  2016-01-01 01:00:00         0.0000
1           6          745  2016-01-01 02:00:00         0.0000
2           6          745  2016-01-01 03:00:00         0.0000
3           6          745  2016-01-01 04:00:00         0.0000
4           6          745  2016-01-01 05:00:00         0.0000
...       ...          ...                  ...            ...
3372        6          745  2016-12-31 13:00:00         2.9249
3373        6          745  2016-12-31 19:00:00         9.8624
3374        6          745  2016-12-31 21:00:00        27.0749
3375        6          745  2016-12-31 22:00:00         1.4999
3376        6          745  2016-12-31 23:00:00         1.4999

[3377 rows x 4 columns]

In [19]:
# create empty DataFrame, 
# I am going to put the number of zeros and indexes with non zero for each building into this DF
building_non_zero = pd.DataFrame(columns = ['building_id','zeros','start_from','non_zero_idx_start','non_zero_idx_end'])
building_non_zero

Empty DataFrame
Columns: [building_id, zeros, start_from, non_zero_idx_start, non_zero_idx_end]
Index: []

below shows the buildings belong to site_id=0

In [66]:
building_prepared_df.query('site_id==0').building_id.unique()

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104], dtype=int64)

below two queries show after building_id > 104, there are very few building starting with 0 meter_reading for meter0. ---- This means something wrong on the meter 0 meter_reading data for most of buildings in site_id = 0 (also see the query after)

In [67]:
np.cumsum(building_non_zero.iloc[105:].zeros != 0)

105      0
106      0
107      0
108      0
109      0
        ..
1408    21
1409    21
1410    22
1411    22
1412    22
Name: zeros, Length: 1308, dtype: int32

In [68]:
np.cumsum(building_non_zero.iloc[:105].zeros != 0)

0        1
1        2
2        3
3        4
4        5
      ... 
100    100
101    101
102    102
103    103
104    104
Name: zeros, Length: 105, dtype: int32

I will define a method to clean up exceptional zeros

In [163]:
# meterlist = [train_set0,train_set1,train_set2,train_set3]

site_id  building_id            timestamp  meter_reading
0               0            0  2016-01-01 00:00:00          0.000
1               0            0  2016-01-01 01:00:00          0.000
2               0            0  2016-01-01 02:00:00          0.000
3               0            0  2016-01-01 03:00:00          0.000
4               0            0  2016-01-01 04:00:00          0.000
...           ...          ...                  ...            ...
12017403       15         1448  2016-12-31 19:00:00          3.400
12017404       15         1448  2016-12-31 20:00:00          2.700
12017405       15         1448  2016-12-31 21:00:00          3.050
12017406       15         1448  2016-12-31 22:00:00          3.125
12017407       15         1448  2016-12-31 23:00:00          2.850

[12017408 rows x 4 columns]

In [20]:
## boolean_all = np.array([]).reshape(-1,1).astype(bool)

def excep_zero_cleanup (meterdata):
    boolean_nonzeros = np.ones((meterdata.shape[0],1), dtype=bool)

    for each in meterdata.building_id.unique():
        query_temp = 'building_id == ' + str(each)
        build_data = meterdata.query(query_temp)
        build_start = build_data.head(1).index[0]
        index_firstnon0 = build_start
        cumsum_non0_rowcount = np.cumsum(build_data.meter_reading != 0)
        counts = 1

            # not only just check first non-0, but also check the next two rows, 
            # if they are 0, keep looking for the next non-0 
        while True:
            if cumsum_non0_rowcount.tail(1).values[0] < counts:
                index_firstnon0 = build_data.tail(1).index[0]+1
                break
            index_firstnon0= build_data[(cumsum_non0_rowcount == counts)].head(1).index[0]
            if meterdata.meter_reading[(index_firstnon0+1)] > 0 and meterdata.meter_reading[(index_firstnon0+2)] > 0:
                break
            counts+=1
        boolean_nonzeros[build_start:index_firstnon0] = False

    # filter rows with zero:
    return meterdata[boolean_nonzeros].copy()

In [21]:
train_set0.shape, train_set1.shape, train_set2.shape, train_set3.shape

((12017408, 4), (4160093, 4), (2686678, 4), (1261426, 4))

In [22]:
train_set0 = excep_zero_cleanup(train_set0)
train_set1 = excep_zero_cleanup(train_set1)
train_set2 = excep_zero_cleanup(train_set2)
train_set3 = excep_zero_cleanup(train_set3)

In [224]:
train_set0.shape, train_set1.shape, train_set2.shape, train_set3.shape

((11633169, 4), (3944467, 4), (2658513, 4), (1250623, 4))

done with execptional zero meter_reading deletion

In [23]:
# man..... there are still lots of zeros, not on the begining of the year, but in the end (or middle) of the year.
# what should we do with them? just delete all of them????, crazy....
train_set0[train_set0.meter_reading == 0].groupby('building_id').count().sort_values(by='meter_reading',ascending=False).head(50)

site_id  timestamp  meter_reading
building_id                                   
1345            6293       6293           6293
857             5561       5561           5561
853             5304       5304           5304
817             4439       4439           4439
815             4248       4248           4248
886             3585       3585           3585
1446            3575       3575           3575
590             3563       3563           3563
1264            3199       3199           3199
1241            3147       3147           3147
260             2559       2559           2559
803             2320       2320           2320
258             2282       2282           2282
1227            2170       2170           2170
1234            2014       2014           2014
1233            2014       2014           2014
1359            1982       1982           1982
180             1962       1962           1962
48              1940       1940           1940
46              1922       1922           1922
848             1655       1655           1655
53              1539       1539           1539
994             1446       1446           1446
1303            1437       1437           1437
809             1416       1416           1416
12              1243       1243           1243
845             1200       1200           1200
737             1051       1051           1051
1319            1009       1009           1009
799              952        952            952
943              851        851            851
1314             828        828            828
186              759        759            759
1275             741        741            741
1276             741        741            741
1272             740        740            740
1280             724        724            724
1250             710        710            710
1225             700        700            700
1226             700        700            700
1302             692        692            692
1300             688        688            688
1279             685        685            685
1278             685        685            685
1277             685        685            685
1294             675        675            675
1299             673        673            673
1297             673        673            673
1296             673        673            673
1293             666        666            666

In [24]:
train_set0.query('building_id == 1345').tail(50)

site_id  building_id            timestamp  meter_reading
11347195       15         1345  2016-12-29 22:00:00            0.0
11347196       15         1345  2016-12-29 23:00:00            0.0
11347197       15         1345  2016-12-30 00:00:00            0.0
11347198       15         1345  2016-12-30 01:00:00            0.0
11347199       15         1345  2016-12-30 02:00:00            0.0
11347200       15         1345  2016-12-30 03:00:00            0.0
11347201       15         1345  2016-12-30 04:00:00            0.0
11347202       15         1345  2016-12-30 05:00:00            0.0
11347203       15         1345  2016-12-30 06:00:00            0.0
11347204       15         1345  2016-12-30 07:00:00            0.0
11347205       15         1345  2016-12-30 08:00:00            0.0
11347206       15         1345  2016-12-30 09:00:00            0.0
11347207       15         1345  2016-12-30 10:00:00            0.0
11347208       15         1345  2016-12-30 11:00:00            0.0
11347209       15         1345  2016-12-30 12:00:00            0.0
11347210       15         1345  2016-12-30 13:00:00            0.0
11347211       15         1345  2016-12-30 14:00:00            0.0
11347212       15         1345  2016-12-30 15:00:00            0.0
11347213       15         1345  2016-12-30 16:00:00            0.0
11347214       15         1345  2016-12-30 17:00:00            0.0
11347215       15         1345  2016-12-30 18:00:00            0.0
11347216       15         1345  2016-12-30 19:00:00            0.0
11347217       15         1345  2016-12-30 20:00:00            0.0
11347218       15         1345  2016-12-30 21:00:00            0.0
11347219       15         1345  2016-12-30 22:00:00            0.0
11347220       15         1345  2016-12-30 23:00:00            0.0
11347221       15         1345  2016-12-31 00:00:00            0.0
11347222       15         1345  2016-12-31 01:00:00            0.0
11347223       15         1345  2016-12-31 02:00:00            0.0
11347224       15         1345  2016-12-31 03:00:00            0.0
11347225       15         1345  2016-12-31 04:00:00            0.0
11347226       15         1345  2016-12-31 05:00:00            0.0
11347227       15         1345  2016-12-31 06:00:00            0.0
11347228       15         1345  2016-12-31 07:00:00            0.0
11347229       15         1345  2016-12-31 08:00:00            0.0
11347230       15         1345  2016-12-31 09:00:00            0.0
11347231       15         1345  2016-12-31 10:00:00            0.0
11347232       15         1345  2016-12-31 11:00:00            0.0
11347233       15         1345  2016-12-31 12:00:00            0.0
11347234       15         1345  2016-12-31 13:00:00            0.0
11347235       15         1345  2016-12-31 14:00:00            0.0
11347236       15         1345  2016-12-31 15:00:00            0.0
11347237       15         1345  2016-12-31 16:00:00            0.0
11347238       15         1345  2016-12-31 17:00:00            0.0
11347239       15         1345  2016-12-31 18:00:00            0.0
11347240       15         1345  2016-12-31 19:00:00            0.0
11347241       15         1345  2016-12-31 20:00:00            0.0
11347242       15         1345  2016-12-31 21:00:00            0.0
11347243       15         1345  2016-12-31 22:00:00            0.0
11347244       15         1345  2016-12-31 23:00:00            0.0

In [25]:
# decision made: try deleting all zeros and put them in the model first, to see the score
train_set0_no0 = train_set0[train_set0.meter_reading != 0 ].copy()
train_set0_no0

site_id  building_id            timestamp  meter_reading
3378            0            0  2016-05-20 18:00:00     232.753006
3379            0            0  2016-05-20 19:00:00     277.802002
3380            0            0  2016-05-20 20:00:00     238.895996
3381            0            0  2016-05-20 21:00:00     240.944000
3382            0            0  2016-05-20 22:00:00     252.546997
...           ...          ...                  ...            ...
12017403       15         1448  2016-12-31 19:00:00       3.400000
12017404       15         1448  2016-12-31 20:00:00       2.700000
12017405       15         1448  2016-12-31 21:00:00       3.050000
12017406       15         1448  2016-12-31 22:00:00       3.125000
12017407       15         1448  2016-12-31 23:00:00       2.850000

[11486383 rows x 4 columns]

In [26]:
# Multiply by 0.2931 to get to model inputs into kWh like the other sites, 
# and 3.4118 to get back to kBTU for scoring
converted_meter0 = train_set0_no0[ train_set0_no0.site_id ==0].meter_reading.mul(0.2931)
converted_meter0

3378       68.219906
3379       81.423767
3380       70.020416
3381       70.620686
3382       74.021525
             ...    
908404    116.533923
908405    114.333330
908406    119.935057
908407    120.735214
908408     80.123279
Name: meter_reading, Length: 560186, dtype: float64

In [32]:
# keep 3 digits 
converted_meter0 = converted_meter0.round(3)
converted_meter0

3378       68.220
3379       81.424
3380       70.020
3381       70.621
3382       74.022
           ...   
908404    116.534
908405    114.333
908406    119.935
908407    120.735
908408     80.123
Name: meter_reading, Length: 560186, dtype: float64

In [33]:
# put the change back to the column
train_set0_no0.update(converted_meter0)
train_set0_no0

site_id  building_id            timestamp  meter_reading
3378            0            0  2016-05-20 18:00:00         68.220
3379            0            0  2016-05-20 19:00:00         81.424
3380            0            0  2016-05-20 20:00:00         70.020
3381            0            0  2016-05-20 21:00:00         70.621
3382            0            0  2016-05-20 22:00:00         74.022
...           ...          ...                  ...            ...
12017403       15         1448  2016-12-31 19:00:00          3.400
12017404       15         1448  2016-12-31 20:00:00          2.700
12017405       15         1448  2016-12-31 21:00:00          3.050
12017406       15         1448  2016-12-31 22:00:00          3.125
12017407       15         1448  2016-12-31 23:00:00          2.850

[11486383 rows x 4 columns]

In [34]:
train_set0 = reduce_mem_usage(train_set0)
train_set1 = reduce_mem_usage(train_set1)
train_set2 = reduce_mem_usage(train_set2)
train_set3 = reduce_mem_usage(train_set3)
train_set0_no0 = reduce_mem_usage(train_set0_no0)

Mem. usage decreased from 255.17 Mb to 255.17 Mb (0.0% reduction)
Mem. usage decreased from 86.52 Mb to 86.52 Mb (0.0% reduction)
Mem. usage decreased from 58.31 Mb to 58.31 Mb (0.0% reduction)
Mem. usage decreased from 27.43 Mb to 27.43 Mb (0.0% reduction)
Mem. usage decreased from 295.77 Mb to 251.95 Mb (14.8% reduction)


### save results

In [35]:
building_prepared_df.to_csv('C:/data/Kaggle_ASHRAE_energy/building_prepared_df.csv', index = False)
weather_train_prepared_df.to_csv('C:/data/Kaggle_ASHRAE_energy/weather_train_prepared_df.csv', index = False)

In [36]:
train_set0.to_csv('C:/data/Kaggle_ASHRAE_energy/train_set0_deleteheadzeros.csv', index = False)
train_set1.to_csv('C:/data/Kaggle_ASHRAE_energy/train_set1_deleteheadzeros.csv', index = False)
train_set2.to_csv('C:/data/Kaggle_ASHRAE_energy/train_set2_deleteheadzeros.csv', index = False)
train_set3.to_csv('C:/data/Kaggle_ASHRAE_energy/train_set3_deleteheadzeros.csv', index = False)

In [37]:
train_set0_no0.to_csv('C:/data/Kaggle_ASHRAE_energy/train_set0_no0.csv', index = False)